In [1]:
# установить свежую версию
!pip install scipy==1.5.2

     |████████████████████████████████| 25.9MB 168kB/s 
ERROR: tensorflow 2.3.0 has requirement scipy==1.4.1, but you'll have scipy 1.5.2 which is incompatible.
ERROR: albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.
  Found existing installation: scipy 1.4.1
    Uninstalling scipy-1.4.1:
      Successfully uninstalled scipy-1.4.1


In [2]:
import math
import numpy as np
import pandas as pd
from scipy import stats as st
import matplotlib.pyplot as plt 

#### Коэффициент корреляции

Рассмотрим случайный вектор $(\xi, \eta)$. 

Мерой взаимосвязи случайных величин является коэффициент корреляции

$p(\xi,eta) = \frac{cov(\xi, \eta)}{\sqrt{D\xi} \sqrt{D\eta}}$

Коэф. корреляции характеризует степень линейной взаимосвязи между сл. величинами $\xi$ и $\eta$.

Если $\xi$ и $\eta$ независимы $\Rightarrow p(\xi, \eta) = 0$ 

обратное не верно!

$|p| = 1 \Rightarrow \eta = a + b\xi$

Если $p < 0$, то говорят об отрицательной корреляцией.

Если $p > 0$, то говорят об положительной корреляцией. Полож. коррел. означает, что если мы наблюдаем увеличение значений $\xi$, то мы будем наблюдать увеличение $\eta$.

Статистическая зависимость двух сл. величин не всегда означает, что одна сл. величина влияет на другую сл. вел. мы просто наблюдаем взаимосвязь их значений! На обе сл. вел. может влиять некоторая третья сл. вел., кот-я пораждает зависимость значений.

Наша задача, по наблюдаемой выборки, построить оценку коэф. корреляции и выяснить, зависимы или не зависимы сл. вел.

$(X, Y)_{[n]} = (x_1, y_1), ..., (x_n, y_n)$

#### Выборочный коэф. корреляции Пирсона

Мера линейной взаимосвязи.

$r_{X,Y} = \frac{\frac{1}{n} \sum \limits_{i=1}^n (x_i - \overline X)(y_i - \overline Y) }{ \sqrt{D^*_X D^*_Y}}$, где

$D^*_X = \frac{1}{n} \sum \limits_{i=1}^n (x_i - \overline X)^2$

$D^*_Y = \frac{1}{n} \sum \limits_{i=1}^n (y_i - \overline Y)^2$

Коэф. кор. Пирсона имеет смысл использовать в том случае, когда наблюдаемые сл. величины измерены в шкале отношений.



**Шкалой отношений** называют такую шкалу с непрерывным множеством числовых значений, в которой о двух сопоставляемых объектах можно сказать не только, одинаковы они или различны, не только, в какой из них признак выражен сильнее, но и во сколько раз сильнее этот признак выражен.

Коэф. кор. Пирсона, хорошо отражает наличие линейной взаимосвязи между двумя сл. величинами, в случае, когда они имеют совместное нормальное распределение или распределение близкое к нормальному.

##### Scipy

In [6]:
X = np.array([3.25, 2.51, 1.46, 2.37, 2.45, -0.07, 3.78, 0.53, 2.09, 1.74])
Y = np.array([0.62, 0.95, 0.37, 1.64, 0.60, 0.19, 1.81, 0.24, 1.26, -0.93]) 

r, p_value = st.pearsonr(X, Y)
r

0.5621846665498427

##### Слуйчай совместного нормального распределения

Пусть случайный вектор имеет совместное нормальное распределение $(\xi, \eta) \sim N(a, K)$, где

$a = (a_1, a_2)$

$K = \begin{pmatrix}
  \sigma^2_1 & \sigma_1 \sigma_2 p \\
  \sigma_1 \sigma_2 p & \sigma^2_2
\end{pmatrix}$ ковариационная матрица, на диагонали стоят дисперсии сл. величин.

$\xi$ и $\eta$ независимы $\Leftrightarrow p(\xi, \eta) = 0$

Будем предполагать, что выборка $(X, Y)_{[n]}$ получена из двумерного, нормально распределенного сл. вектора.

##### Приближенный дов. интер. для $p$

Введем статистику

$Z = \frac{1}{2} ln \frac{1 + r}{1 - r}$, где

$r$ коэф. корреляции, предполагаем, что $|r| \neq 1$



При количестве наблюдений $n > 50$, мы можем считать, что распределение статистики $Z$  является нормальным.

$EZ = \frac{1}{2} ln \frac{1 + p}{1 - p}$

$DZ = \frac{1}{n-3}$

$Z = N(EZ, \sqrt{DZ})$

Зададимся уровнем доверия $1 - \alpha$. Найдём квантили **стандартного нормального распределения** $u_{\frac{\alpha}{2}}$ и $u_{1 - \frac{\alpha}{2}}$.

$P(u_{\frac{\alpha}{2}}$ < $\frac{Z - EZ}{\sqrt{DZ}}$ < $u_{1 - \frac{\alpha}{2}}) = 1 - \alpha$

Подставим мат. ож. и дисперсию, преобразуем и получим интервал

$th(\frac{1}{2} ln \frac{1 + r}{1 - r} - \frac{u_{1 - \frac{\alpha}{2}}}{\sqrt{n-3}} < p < th(\frac{1}{2} ln \frac{1 + r}{1 - r} + \frac{u_{1 - \frac{\alpha}{2}}}{\sqrt{n-3}})$, где

$th$ гиперболический тангенс

##### Проверка гипотезы $p = p_0$

$H_0: p = p_0$ 

коэф. корреляции равен какому-то значению $p_0$

Возьмем статистику 

$Z = \frac{Arth(r_{X, Y}) - Arth(p_0)}{\frac{1}{\sqrt{n-3}}} \sim N(0, 1)$, где

$Arth$ обратный гиперболический тангенс.

$r_{X, Y}$ выборочный коэф. корреляции Пирсона


##### Критическая область

$H_1: p > p_0 \;\; V_k: (u_{1 - \alpha}, + \infty)$

$H_1: p < p_0 \;\; V_k: (- \infty, u_{\alpha})$

$H_1: p \neq p_0 \;\; V_k: (- \infty, u_{\frac{\alpha}{2}}) \cup (u_{1 - \frac{\alpha}{2}}, + \infty)$

$Z_в \in V_k \Rightarrow H_0$ отклоняется.

##### Проверка гипотезы $p = 0$

Мы предполагаем, что выборка получена из нормально распр. ген. совокупности.

Статистика 

$t = \frac{r_{X, Y} \sqrt{n-2}}{\sqrt{1 - r^2_{X, Y}}} \sim T_{n-2}$, где

$T_{n-2}$ распр. Стьюдента.

$r_{X, Y}$ выборочный коэф. корреляции Пирсона

##### Критическая область

$H_1: p > 0 \;\; V_k (t_{1 - \alpha, n-2}, + \infty)$ положительная корреляция

$H_1: p < 0 \;\; V_k (- \infty, t_{\alpha, n-2})$ отрицательная корреляция

$H_1: p \neq 0 \;\; V_k (- \infty, t_{\frac{\alpha}{2}, n-2}) \cup (t_{1 - \frac{\alpha}{2}, n-2}, + \infty)$ имеется взаимосвязь

$t_в \in V_k \Rightarrow H_0$ отклоняется.



---



##### Задача №1

По выборке из двумерного нормального распределения, найдите выборочный коэффициент корреляции Пирсона.

Зависимы ли компоненты случайного вектора? Проверьте гипотезу с помощью точного критерия. Уровень значимости 0.05

Найдите значение статистики критерия и примите статистическое решение.

**Решение:**



По выборке из двумерного нормального распределения, найдите выборочный коэффициент корреляции Пирсона.

$r_{X,Y} = \frac{\frac{1}{n} \sum \limits_{i=1}^n (x_i - \overline X)(y_i - \overline Y) }{ \sqrt{D^*_X D^*_Y}}$, где

$D^*_X = \frac{1}{n} \sum \limits_{i=1}^n (x_i - \overline X)^2$

$D^*_Y = \frac{1}{n} \sum \limits_{i=1}^n (y_i - \overline Y)^2$

In [8]:
X = np.array([3.25, 2.51, 1.46, 2.37, 2.45, -0.07, 3.78, 0.53, 2.09, 1.74])
Y = np.array([0.62, 0.95, 0.37, 1.64, 0.60, 0.19, 1.81, 0.24, 1.26, -0.93]) 

r, p_value = st.pearsonr(X, Y)
r

0.5621846665498427

Зависимы ли компоненты случайного вектора? Проверьте гипотезу с помощью точного критерия. Уровень значимости 0.05

Проверяем гипотезу

$H_0: p = 0$

$H_1: p \neq 0$

$t = \frac{r_{X, Y} \sqrt{n-2}}{\sqrt{1 - r^2_{X, Y}}} \sim T_{n-2}$, где

$T_{n-2}$ распр. Стьюдента.

$r_{X, Y}$ выборочный коэф. корреляции Пирсона

In [9]:
n = len(X)

tv = r * np.sqrt(n - 2) / np.sqrt(1 - r**2)
tv

1.9227035159310693

$V_k = (- \infty, t_{\frac{\alpha}{2}, n-2}) \cup (t_{1 - \frac{\alpha}{2}, n-2}, + \infty)$

In [10]:
alpha = 0.05
ls, rs = st.t.ppf(alpha/2, n-2), st.t.ppf(1 - (alpha/2), n-2)
ls, rs

(-2.306004135033371, 2.3060041350333704)

$V_k = (- \infty, -2.3) \cup (2.3< + \infty)$

$t_в = 1.92$

$t_в \notin V_k \Rightarrow H_0$ не отвергается. Сл. вел. независимы.

**Ответ:**



$r_{X,Y} = 0.56$ коэф. Пирсона.

$t_в = 1.92$ выборочная статистика.

$t_в \notin V_k \Rightarrow H_0$ не отвергается. Сл. вел. независимы.

#### Порядковые шкалы

Признаки и характеристики объектов не всегда возможно измерить в шкале отношений.  Однако и в этом случае нам необходимо уметь проверять независимость признаков.

Мы уже знакомы с применением критерия хи-квадрат для проверки гипотезы о независимости категориальных признаков.

В случае, если признаки измерины в порядковой шкале, то для проверки независимости признаков, мы можем воспользоваться ранговыми коэф. корреляции:

* коэф. ранговой корреляции Спирмена
* коэф. ранговой корреляции Кенделла

**Порядками или ординальными шкалами**, наз-ся шкалы, в которых существенен лишь взаимный порядок в котором следуют результаты измерений, а не их количественные значения.

Пример, баллы за экзамен или результаты соц. опроса.

Рассмотрим объекты 

$O_1, ..., O_n$

Каждый из объектов обладает признаком $A$ и $B: (x_i, y_i), 1 \leq i \leq n$

каждому объекту соотвествует пара $(x_i, y_i)$, $x_i$ это значение соответствующее признаку $A$, $y_i$ это значение соответствующее признаку $B$.

##### Коэф. ранговой корреляции Спирмена

###### На чём основан

В основе лежит идея выборочного коэф. корреляции, только в данном случае, вместо самих значений наблюдений, мы будем использовать ранги.




Проранжируем наблюдения

$x_1, ..., x_n \Rightarrow r_1, ..., r_n$

$y_1, ..., y_n \Rightarrow s_1, ..., s_n$, где

$r_i$ и $s_i$ последовательность рангов.

Для каждого объекта $O_i$ наблюдения заменим на пару рангов

$(x_i, y_i) \Rightarrow (r_i, s_i)$

ранг - это номер элемента в вариационном ряду.

###### Коэффициент

Мера монотонной взаимосвязи.

Подставим в формулу вычисления значения корреляции, вместо $(x_i, y_i)$ их ранги

$r_S = \frac{\frac{1}{n} \sum \limits_{i=1}^n (r_i - \overline r_i)(s_i - \overline s_i)}{\sqrt{ \frac{1}{n} \sum \limits_{i=1}^n (r_i - \overline r_i)^2 \frac{1}{n} \sum \limits_{i=1}^n (s_i - \overline s_i)^2 }} = 1 - \frac{6 \sum \limits_{i=1}^n (s_i - r_i)^2}{n^3 - n}$, где

$r_S$ коэф. ранговой корреляции Спирмена.

$r_S \in [-1, 1]$

$r_S = 0$ значит последовательности независимы



###### Статистика

В случае, если объем выборки велик и признаки А и В независимы, статистика имеет распред. Мы можем использовать данное приближенное распределение для проверки гипотезы и построения критической области.

$\sqrt{n-1} r_S \sim N(0, 1)$

###### Пример

Предположим, что двух экспертов попросили отранжировать пять видов вин, по их вкусовым качествам. Всего вин a, b, c, d, e. Отранжировали от лучшего к худшему.

эксперт 1: a, b, d, e, c

эксперт 2: b, a, e, c, d

Узнаем, коррелируют ли ранжировки экспертов.

Проранжируем наблюдения

эксперт 1:

a, b, d, e, c

1, 2, 3, 4, 5

эксперт 2:

b, a, e, c, d

1, 2, 3, 4, 5



Найдём вначале сумму квадратов разницы рангов

$S = \sum \limits_{i=1}^n(r_i - s_i)^2$

Вину a у первого ранг 1 у второго 2

$S = \sum \limits_{i=1}^n(r_i - s_i)^2 = (1 - 2)^2 + (2 - 1)^2 + (5 - 4)^2 + (3 - 5)^2 + (4 - 3)^2 = 8$




Коэф. ранговой кор. Спирмена

$r_S = 1 - \frac{6 \cdot 8}{5^3 - 5} = 0.52$

###### Проверка гипотезы

Проверим гипотезу о независимости признаков

$H_0:$ признаки $A$ и $B$ взаимно независимы.

$u_{\alpha}$ квантиль стандартного нормального распределения

$H_1: r_S > 0$, таблицы: $(c_1, 1]$, аппроксимация: $(u_{1 - \alpha}, + \infty)$

$H_1: r_S < 0$, таблицы: $[-1, c_2]$, аппроксимация: $(- \infty, u_{\alpha})$

$H_1: r_S \neq 0$, таблицы: $[-1, c_3] \cup (c_4, 1]$, аппроксимация: $(-\infty, u_{\frac{\alpha}{2}}) \cup (u_{1 - \frac{\alpha}{2}}, + \infty)$

$r_{S} \in V_k \Rightarrow H_0$ отвергается.

In [11]:
alpha = 0.05
ls = st.norm.ppf(alpha/2)
rs = st.norm.ppf(1 - (alpha/2))
ls, rs

(-1.9599639845400545, 1.959963984540054)

$r_{Sв} = 0.52$

$V_k = (- \infty, -1.95) \cup (1.95, + \infty)$

$r_{Sв} \notin V_k \Rightarrow H_0$ не отвергается. Значит между сл. величинами есть связь.

###### Задача №1

С помощью двух алгоритмов ранжирования были отранжированы 10 документов:

d1,d2,d3,d4,d5,d6,d7,d8,d9,d10

Ранжировки:

1) d6 d2 d10 d8 d3 d9 d1 d5 d4 d7

2) d10 d6 d8 d2 d3 d9 d1 d5 d7 d4


Насколько сильно отличаются ранжировки данных алгоритмов?

Чтобы ответить на вопрос, найдите коэффициент ранговой корреляции Спирмена

**Решение:**

Если мы вместо $di$ будем записывать только $i$, то получим номера выборки. Они идут от одного до десяти, если мы составим 

In [22]:
X = [6, 2, 10, 8, 3, 9, 1, 5, 4, 7]
Y = [10, 6, 8, 2, 3, 9, 1, 5, 7, 4]

Вычислим сумму квадратов разницы рангов

$S = \sum \limits_{i=1}^n(r_i - s_i)^2$

In [29]:
S = np.sum([(X.index(i) - Y.index(i))**2 for i in range(1, 10)])
S

8

Вычислим коэф. ранговой корреляции Спирмена

$r_S = 1 - \frac{6 \cdot S}{\sqrt{n^3 - n}}$

In [32]:
n = len(X)
rS = 1 - ( (6 * S) / (n**3 - n) )
rS

0.9515151515151515

**Ответ:**

$r_S = 0.95$

##### Коф. ранговой корреляции Кенделла

###### На чём основан

Основа на сравнении доли согласованых пар и несогласованных.

Рассмотрим объекты 

$O_1, ..., O_n$

Каждый из объектов обладает признаком $A$ и $B: (x_i, y_i), 1 \leq i \leq n$

каждому объекту соотвествует пара $(x_i, y_i)$, $x_i$ это значение соответствующее признаку $A$, $y_i$ это значение соответствующее признаку $B$.

Проранжируем наблюдения

$x_1, ..., x_n \Rightarrow r_1, ..., r_n$

$y_1, ..., y_n \Rightarrow s_1, ..., s_n$, где

$r_i$ и $s_i$ последовательность рангов.

Рассмотрим два объекта

$O_i$ и соответствующее наблюдение $(x_i, y_i)$

$O_j$ и соответствующее наблюдение $(x_j, y_j)$

Предположим, что все ранги различны. Назовём пару наблюдений согласованной, если порядок следования рангов обоих признаков совпадает. А именно, если выполняется одно из двух условий:

*  $r_i > r_j, s_i > r_j$

*  $r_i < r_j, s_i < r_j$

Обозначим через

$C$ число согласованныx пар.

$D$ число не согласованныx пар.

###### Коэффициент

Мера монотонной взаимосвязи.

$r_K = \frac{C - D}{C + D} = \frac{2 (C - D)}{n(n-1)}$

$ \frac{n(n-1)}{2}$ общее число пар

Если порядок следования рангов признаков для всех пар наблюдений совпадает, в этом случае коэф примет значение 

$r_K = 1$

если порядок строго противоположный

$r_K = -1$

если равен нулю

$r_K = 0$ значит независимы признаки $А$ и $В$

$r_K \in [-1, 1]$

###### Статистика

В случае, если объем выборки велик и признаки А и В независимы, статистика имеет распред. Мы можем использовать данное приближенное распределение для проверки гипотезы и построения критической области.

$\sqrt{ \frac{9n(n+1)}{2(2n+5)} } r_K \sim N(0, 1)$

###### Пример

Предположим, что двух экспертов попросили отранжировать пять видов вин, по их вкусовым качествам. Всего вин a, b, c, d, e. Отранжировали от лучшего к худшему.

эксперт 1: a, b, d, e, c

эксперт 2: b, a, e, c, d

Узнаем, коррелируют ли ранжировки экспертов.

Выпишем все возможные пары вин в том порядке в котором они отранжировали их.

эксперт 1: (a, b), (a, d), (a, c), (b, d), (b, e), (b, c), (d, e), (d, c), (e, c)

эксперт 2: (b, a), (b, e), (b, c), (b, d), (a, e), (a, c), (a, a), (e, c), (e, d), (c, d)

Найдём все согласованные пары, т.е. те в которых порядок вин совпадает.

(a, d), (a, c), (b, d), (b, e), (b, c), (e, c) согласованные пары (есть у обоих экспертов)

(a, b), (d, e), (d, c), (b, a), (e, d), (c, d) несогласованные пары

$C = 7$ согласованные пары

$D = 3$ несогласованные пары

$r_k = \frac{7 - 3}{3 + 3} = 0.1$

###### Проверка гипотезы

$H_0:$ признаки $A$ и $B$ взаимно независимы.

$H_1: r_S > 0$, таблицы: $(c_1, 1]$, аппроксимация: $(u_{1 - \alpha}, + \infty)$

$H_1: r_S < 0$, таблицы: $[-1, c_2]$, аппроксимация: $(- \infty, u_{\alpha})$

$H_1: r_S \neq 0$, таблицы: $[-1, c_3] \cup (c_4, 1]$, аппроксимация: $(-\infty, u_{\frac{\alpha}{2}}) \cup (u_{1 - \frac{\alpha}{2}}, + \infty)$

In [ ]:
$r_{K} \in V_k \Rightarrow H_0$ отвергается.

###### Задача №1

С помощью двух алгоритмов ранжирования были отранжированы 10 документов.

d1,d2,d3,d4,d5,d6,d7,d8,d9,d10

Насколько сильно отличаются ранжировки данных алгоритмов?
1) d6 d2 d10 d8 d3 d9 d1 d5 d4 d7

2) d10 d6 d8 d2 d3 d9 d1 d5 d7 d4

Чтобы ответить на вопрос,  найдите коэффициент ранговой корреляции Кенделла

**Решение:**



In [35]:
# ранжировки
x1 = [6, 2, 10, 8, 3, 9, 1, 5, 4, 7]
x2 = [10, 6, 8, 2, 3, 9, 1, 5, 7, 4]
n = len(x1)
n_paips = (n-1)*n/2 # количество пар

# функция вычисления всех пар ...
pairs = lambda x: [(x[i], x[j]) for i in range(len(x)-1) 
                                    for j in range(i+1,len(x))]

# вычисляем пары
ps1 = pairs(x1)
ps2 = pairs(x2)

# вычисляем кол-во согласованных и несогласованных ...
C = len(set(ps1).intersection(set(ps2)))
D = n_paips - C

$r_K = \frac{C - D}{C + D}$

In [39]:
rK = (C - D) / (C + D)
rK

0.8222222222222222

**Ответ:**

$r_{Kв} = 0.8$